text/0000-signals-in-lwc-templates.md

```diff
@@ -0,0 +1,196 @@
+1. __Complexity and Overhead__: The use of a membrane based on proxies introduces additional complexity and overhead in component design and state management, from read-and-write to read-only objects, it can be challenging for developers to debug and reason about data flow and updates.
+1. __Performance Constraints__: The current model relies on components to re-render fully to reflect state changes, which can be inefficient, especially for large and complex component trees.
+
+By adopting a Signals-based reactivity in templates, LWC can overcome these challenges:
+
+1. __Framework Agnosticism__: With Signals, LWC components can interact with any external data store that follows the protocol. This change would make LWC more agnostic and flexible, allowing for easier integration with various state management patterns and libraries.
+1. __Simplified State Management__: Signals provide a straightforward way to manage reactive data. By using signals, developers can reduce complexity, making it easier to write and maintain component logic.
+1. __Performance Optimizations__: The proposed model opens up possibilities for performance optimizations by allowing selective rehydration of the UI. Components could update the DOM directly in response to state changes without a full re-render, leading to faster updates and reduced resource consumption. This has been a goal of the LWC team for a long time, small incremental improvements has been made, but this effort could accelerate the work in this area.
+
+The motivation for this RFC is to enable LWS, and developers in general, to bring stores and signal-like objects as first class citizens into the LWC ecosystem without introducing significant overhead. To improve the ergonomics of signals in LWC Template, to align LWC with more modern JavaScript practices, simplify component development, and improve performance, making LWC a developers friendly.
+
+## Detailed Design
+
+The detailed design of the new reactivity model revolves around the concept of Signals, which are reactive primitives that allow for fine-grained control over state management and UI updates. Here's how it will look in practice:
+
+### Reactivity with Signals
+
+A Signal is an object that holds a value and allows components to react to changes to that value. It exposes a `.value` property for accessing the current value, and `.subscribe` methods for responding to changes.
+
+```javascript
+import { signal } from 'some/signals';
+
+export default class ExampleComponent extends LightningElement {
+    count = signal(0);
+
+    increment() {
+        this.count.value++;
+    }
+}
+```
+
+In the template, we can bind directly to the `.value` property:
+
+```html
+<template>
+    <button onclick={increment}>Increment</button>
+    <p>{count.value}</p>
+</template>
+```
```

Comment:
What happens if I bind with `count`? 
I really did not like to be forced to use `.value` all the time. I recall you have already answered this question but I could not find your answer.

Reply:
What if lwc provides `ref` and `unref` like vue 3?

https://markus.oberlehner.net/blog/vue-3-composition-api-ref-vs-reactive

Reply:
the problem of not having a hint (e.g.: `.value`), then you have two options:

1. introduce a new syntax, to hint when to use it as a signal.
2. introduce a big overhead to the template runtime execution to test every binding to see if the binding is a signal (I don't think we can afford this).